tests/test_group.py

```diff
@@ -0,0 +1,140 @@
+        job_ids = [job.id for job in group.get_jobs()]
+        jobs = list({as_text(job) for job in self.testconn.smembers(group.key)})
+        self.assertCountEqual(jobs, job_ids)
+        q.empty()
+
+    def test_group_id_added_to_jobs(self):
+        q = Queue(connection=self.testconn)
+        group = Group.create(connection=self.testconn)
+        jobs = group.enqueue_many(q, [self.job_1_data])
+        assert jobs[0].group_id == group.name
+        fetched_job = Job.fetch(jobs[0].id, connection=self.testconn)
+        assert fetched_job.group_id == group.name
+
+    def test_deleted_jobs_removed_from_group(self):
+        q = Queue(connection=self.testconn)
+        group = Group.create(connection=self.testconn)
+        group.enqueue_many(q, [self.job_1_data, self.job_2_data])
+        job = group.get_jobs()[0]
+        job.delete()
+        group.cleanup()
+        redis_jobs = list({as_text(job) for job in self.testconn.smembers(group.key)})
+        assert job.id not in redis_jobs
+        assert job not in group.get_jobs()
+
+    def test_group_added_to_registry(self):
+        q = Queue(connection=self.testconn)
+        group = Group.create(connection=self.testconn)
+        group.enqueue_many(q, [self.job_1_data])
+        redis_groups = {as_text(group) for group in self.testconn.smembers("rq:groups")}
+        assert group.name in redis_groups
+        q.empty()
+
+    @pytest.mark.slow
+    def test_expired_jobs_removed_from_group(self):
+        q = Queue(connection=self.testconn)
+        w = SimpleWorker([q], connection=q.connection)
+        short_lived_job = Queue.prepare_data(say_hello, result_ttl=1)
+        group = Group.create(connection=self.testconn)
+        group.enqueue_many(q, [short_lived_job, self.job_1_data])
+        w.work(burst=True, max_jobs=1)
+        sleep(2)
+        w.run_maintenance_tasks()
+        group.cleanup()
+        assert len(group.get_jobs()) == 1
+        assert self.job_1_data.job_id in [job.id for job in group.get_jobs()]
+        q.empty()
+
+    @pytest.mark.slow
+    def test_empty_group_removed_from_group_list(self):
+        q = Queue(connection=self.testconn)
+        w = SimpleWorker([q], connection=q.connection)
+        short_lived_job = Queue.prepare_data(say_hello, result_ttl=1)
+        group = Group.create(connection=self.testconn)
+        group.enqueue_many(q, [short_lived_job])
+        w.work(burst=True, max_jobs=1)
+        sleep(2)
+        w.run_maintenance_tasks()
+        redis_groups = {as_text(group) for group in self.testconn.smembers("rq:groups")}
+        assert group.name not in redis_groups
+
+    @pytest.mark.slow
+    def test_fetch_expired_group_raises_error(self):
+        q = Queue(connection=self.testconn)
+        w = SimpleWorker([q], connection=q.connection)
+        short_lived_job = Queue.prepare_data(say_hello, result_ttl=1)
+        group = Group.create(connection=self.testconn)
+        group.enqueue_many(q, [short_lived_job])
```

Comment:
You have put a lot of time and effort into this PR which I greatly apppreciate. This is just something extra (not required) but I wonder if we should also have an API from the queue side to make it a bit more consistent with the rest of RQ's APIs:
```python
queue.enqueue(func, group=group)
```

Honestly I have also forgotten whether the earlier iterations had this option.

Reply:
@selwin I believe we did have this at some point and agreed to move it to `batch`
https://github.com/rq/rq/pull/1945#discussion_r1328026775